### PR TITLE
Allow to load a private key from the OpenSSL engine.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Its possible to plugin different behavior into OpenSSL by providing a custom OpenSSL engine. When doing so it's possible to load a private key from the engine directly.

Modifications:

Add new method which allows to load a private key from the OpenSSL engine directly and so can be used later on.

Result:

Be able to implement custom loading / store mechanism of private keys via a custom OpenSSL engine.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@carl-mastrangelo @ejona86 sadly this is again not easily testable :( I tested it locally tho.

File: openssl-dynamic/src/main/c/ssl.c [hunk 1]
```diff
@@ -725,6 +781,19 @@ TCN_IMPLEMENT_CALL(jint, SSL, initialize)(TCN_STDARGS, jstring engine)
             if (ee)
                 ENGINE_free(ee);
         }
+        if (ee != NULL) {
```
Comment (ejona86) [inline openssl-dynamic/src/main/c/ssl.c]: This check seem strange, since `ee` will have been freed and it's possible for `ee != NULL` but `err == APR_ENOTIMPL`. Should this block be moved into the previous block?

File: openssl-dynamic/src/main/c/ssl.c [hunk 2]
```diff
@@ -334,6 +343,47 @@ static long bio_java_bytebuffer_ctrl(BIO* bio, int cmd, long num, void* ptr) {
...
+{
+    const char *password;
+    switch (UI_get_string_type(uis)) {
+    case UIT_PROMPT:
+    case UIT_VERIFY:
+        password = (const char *) UI_get0_user_data(ui);
+        if (password != NULL && (UI_get_input_flags(uis) & UI_INPUT_FLAG_DEFAULT_PWD) != 0) {
+            UI_set_result(ui, uis, password);
+            return 1;
+        }
```
Comment (ejona86) [inline openssl-dynamic/src/main/c/ssl.c]: Add a comment that fallthrough is expected (make it clear there is a missing break on purpose)? Ditto in next function.

File: openssl-dynamic/src/main/c/ssl.c [hunk 3]
```diff
@@ -725,6 +781,19 @@ TCN_IMPLEMENT_CALL(jint, SSL, initialize)(TCN_STDARGS, jstring engine)
             if (ee)
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/ssl.c]: `ee != NULL` while you are here.

**Comment by normanmaurer** [comment]
@ejona86 addressed.

File: openssl-dynamic/src/main/c/ssl.c (hunk 3 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: done

File: openssl-dynamic/src/main/c/ssl.c [hunk 4]
```diff
@@ -334,6 +343,51 @@ static long bio_java_bytebuffer_ctrl(BIO* bio, int cmd, long num, void* ptr) {
...
+ * Supply default password to the engine user interface conversation.
+ * The password is passed by OpenSSL engine from ENGINE_load_private_key()
+ * last argument to the ui and can be obtained by UI_get0_user_data(ui) here.
+ */
+static int ssl_ui_reader(UI *ui, UI_STRING *uis)
+{
+    const char *password;
+    switch (UI_get_string_type(uis)) {
+    case UIT_PROMPT:
+        // fall-through
```
Comment (ejona86) [inline openssl-dynamic/src/main/c/ssl.c]: FWIW, I was fine without a comment here because the fallthrough was obvious and normal.

File: openssl-dynamic/src/main/c/ssl.c [hunk 5]
```diff
@@ -721,10 +781,24 @@ TCN_IMPLEMENT_CALL(jint, SSL, initialize)(TCN_STDARGS, jstring engine)
                 if (!ENGINE_set_default(ee, ENGINE_METHOD_ALL))
                     err = APR_ENOTIMPL;
             }
-            /* Free our "structural" reference. */
-            if (ee)
+            if (ee != NULL) {
```
Comment (ejona86) [inline openssl-dynamic/src/main/c/ssl.c]: The error flow of this is still hard to follow, because it is possible for `err == APR_ENOTIMPL` here. It seems like that isn't on purpose. Although it looks like it would work. Is that on purpose?

File: openssl-dynamic/src/main/c/ssl.c (hunk 4 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: I guess it cant hurt :)

File: openssl-dynamic/src/main/c/ssl.c (hunk 5 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: yeah let me just make it more clear.

**Comment by normanmaurer** [comment]
@ejona86 done :)

**Comment by normanmaurer** [comment]
@ejona86 also fixed a possible memory leak

File: openssl-dynamic/src/main/c/ssl.c [hunk 6]
```diff
@@ -46,6 +51,10 @@ extern apr_pool_t *tcn_global_pool;
 ENGINE *tcn_ssl_engine = NULL;
 void *SSL_temp_keys[SSL_TMP_KEY_MAX];
 
+#ifndef OPENSSL_NO_ENGINE
+UI_METHOD *ui_method = NULL;
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/ssl.c]: should this be defined as `static`?

File: openssl-dynamic/src/main/c/ssl.c (hunk 4 above)
Comment (Scottmitch) [reply openssl-dynamic/src/main/c/ssl.c]: (I'm also fine without fall through here bcz it is expected stacking of cases with no-code ... having the comment brings attention to an area where attention isn't necessary)

File: openssl-dynamic/src/main/c/ssl.c [hunk 7]
```diff
@@ -721,17 +789,29 @@ TCN_IMPLEMENT_CALL(jint, SSL, initialize)(TCN_STDARGS, jstring engine)
                 if (!ENGINE_set_default(ee, ENGINE_METHOD_ALL))
                     err = APR_ENOTIMPL;
             }
-            /* Free our "structural" reference. */
-            if (ee)
-                ENGINE_free(ee);
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/ssl.c]: looks like this was dropped ... are we now leaking `ee`?

File: openssl-dynamic/src/main/c/ssl.c [hunk 8]
```diff
@@ -1711,6 +1791,37 @@ TCN_IMPLEMENT_CALL(void, SSL, setCertificateChainBio)(TCN_STDARGS, jlong ssl,
...
+TCN_IMPLEMENT_CALL(jlong, SSL, loadPrivateKeyFromEngine)(TCN_STDARGS, jstring keyId, jstring password)
+{
+#ifndef OPENSSL_NO_ENGINE
+    char err[ERR_LEN];
+    EVP_PKEY* pkey = NULL;
+
+    TCN_ALLOC_CSTRING(keyId);
+    TCN_ALLOC_CSTRING(password);
+
+    pkey = ENGINE_load_private_key(tcn_ssl_engine, ckeyId, ui_method, (void*) cpassword);
```
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/ssl.c]: `ckeyId` made me do a double take ... magical macros 😕

File: openssl-dynamic/src/main/c/ssl.c (hunk 6 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: Can’t hurt... let me also make tcn_ssl_engine static while I am here

Comment (Scottmitch) [reply openssl-dynamic/src/main/c/ssl.c]: (static will prevent the variable from being visible/exported from this file).

File: openssl-dynamic/src/main/c/ssl.c (hunk 7 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: Nope what was done before was not really correct. We should only call free in the ssl_init_cleanup

Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: So it was moved there

File: openssl-dynamic/src/main/c/ssl.c (hunk 4 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: Ok will remove

File: openssl-dynamic/src/main/c/ssl.c [hunk 9]
```diff
@@ -721,17 +787,29 @@ TCN_IMPLEMENT_CALL(jint, SSL, initialize)(TCN_STDARGS, jstring engine)
...
             }
-            /* Free our "structural" reference. */
-            if (ee)
-                ENGINE_free(ee);
+            if (err == APR_SUCCESS) {
+                // This code is based on libcurl:
+                // https://github.com/curl/curl/blob/curl-7_61_0/lib/vtls/openssl.c#L521
+                ui_method = UI_create_method((char *)"netty-tcnative user interface");
+                if (ui_method != NULL) {
+                    UI_method_set_opener(ui_method, UI_method_get_opener(UI_OpenSSL()));
```
Comment (carl-mastrangelo) [inline openssl-dynamic/src/main/c/ssl.c]: I'm uncomfortable ignoring error codes from these setters.    Could you make them an assert 1 at least?

Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: @carl-mastrangelo sure thing...

Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: that said its actually `0` on success https://www.openssl.org/docs/man1.1.0/crypto/UI_method_set_opener.html

File: openssl-dynamic/src/main/c/ssl.c (hunk 7 above)
Comment (Scottmitch) [reply openssl-dynamic/src/main/c/ssl.c]: can you remove the `ee` intermediate variable then? it introduces the possibility of a leak and makes the ownership less clear.

File: openssl-dynamic/src/main/c/ssl.c (hunk 9 above)
Comment (carl-mastrangelo) [reply openssl-dynamic/src/main/c/ssl.c]: Well, dang, thats scary.  ui.h says the opposite:  https://github.com/openssl/openssl/blob/master/include/openssl/ui.h#L256

```
   All these functions are expected to return:
        0       on error.
        1       on success.
```

Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: I guess this is talking about the "functions" that you use as opener etc itself. not the setter.

**Comment by normanmaurer** [comment]
@carl-mastrangelo PTAL again

File: openssl-dynamic/src/main/java/io/netty/internal/tcnative/SSL.java [hunk 10]
```diff
@@ -598,6 +598,20 @@ public static native boolean setCipherSuites(long ssl, String ciphers)
     public static native void setCertificateBio(
             long ssl, long certBio, long keyBio, String password) throws Exception;
 
+    /**
+     * Load a private key from the used OpenSSL ENGINE via the
+     * <a href="https://www.openssl.org/docs/man1.1.0/crypto/ENGINE_load_private_key.html">ENGINE_load_private_key</a>
+     * function.
+     *
+     * <p>Be sure you understand how OpenSsl will behave with respect to reference counting!
```
Comment (carl-mastrangelo) [inline openssl-dynamic/src/main/java/io/netty/internal/tcnative/SSL.java]: This comment is ominous, I don't really know what I should do about it.  Maybe call `unloadPrivateKey` or something like that?  I don't see that method though.  What should a reader do?

Comment (normanmaurer) [reply openssl-dynamic/src/main/java/io/netty/internal/tcnative/SSL.java]: @carl-mastrangelo added some more docs :)

File: openssl-dynamic/src/main/c/ssl.c (hunk 7 above)
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/ssl.c]: done

**Comment by normanmaurer** [comment]
@Scottmitch also gave ok offline... will merge once the CI pass.
